File: JS/sort/allocationConferenceRoom.js
"use strict";

const fs = require("fs");

let input =
  process.platform === "linux"
    ? fs.readFileSync("/dev/stdin").toString().split("\n")
    : `11
1 4
3 5
0 6
5 7
3 8
5 9
6 10
8 11
8 12
2 13
12 14`.split("\n");

const n = parseInt(input[0]);
const arr = Array(n);
let maxEnd = -1;

for (let i = 1; i < n + 1; i++) {
  const [start, end] = input[i].split(" ").map((v) => parseInt(v));

  if (maxEnd < end) maxEnd = end;

  arr[i - 1] = { start, end };
}

input = null;

arr.sort((a, b) => {
  if (a.end < b.end) return -1;
  else if (a.end > b.end) return 1;
  else {
    return a.start - b.start;
  }
});
let count = 1;
let time = arr[0];

for (let i = 1; i < arr.length; i++) {
  if (time.end <= arr[i].start) {
    time = arr[i];
    count++;
  }
}

console.log(count);

File: src/sort/AllocationConferenceRoom.java
package sort;

import java.io.*;
import java.util.Arrays;

public class AllocationConferenceRoom {
    static class Node implements Comparable<Node>{
        int start;
        int end;

        public Node(int start, int end) {
            this.start = start;
            this.end = end;
        }

        @Override
        public int compareTo(Node o) {
            if(end < o.end){
                return -1;
            }else if(end > o.end){
                return 1;
            }else{
                return start - o.start;
            }
        }
    }
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        int n = Integer.parseInt(br.readLine());
        Node[] arr = new Node[n];
        int maxEnd = -1;

        for (int i = 0; i < n; i++) {
            String[] s = br.readLine().split(" ");
            int start = Integer.parseInt(s[0]), end = Integer.parseInt(s[1]);

            if(maxEnd < end) maxEnd = end;

            arr[i] = new Node(start, end);
        }

        Arrays.sort(arr);

        int count = 1;
        Node time = arr[0];

        for (int i = 1; i < n; i++) {
            if(time.end <= arr[i].start){
                time = arr[i];
                count++;
            }
        }

        bw.write(String.valueOf(count));
        bw.flush();
        bw.close();
    }
}
